programs/mango-v4/src/state/token_conditional_swap.rs

```diff
@@ -0,0 +1,166 @@
+    pub taker_fee_bps: i16,
+
+    /// The maker has to pay premium_price * (1 + maker_fee_bps/10000)
+    pub maker_fee_bps: i16,
+
+    /// indexes of tokens for the swap
+    pub buy_token_index: TokenIndex,
+    pub sell_token_index: TokenIndex,
+
+    pub has_data: u8,
+
+    /// may token purchases create deposits? (often users just want to get out of a borrow)
+    pub allow_creating_deposits: u8,
+    /// may token selling create borrows? (often users just want to get out of a long)
+    pub allow_creating_borrows: u8,
+
+    #[derivative(Debug = "ignore")]
+    pub reserved: [u8; 115],
+}
+
+const_assert_eq!(
+    size_of::<TokenConditionalSwap>(),
+    8 * 6 + 2 * 4 + 2 * 5 + 1 * 3 + 115
+);
+const_assert_eq!(size_of::<TokenConditionalSwap>(), 184);
+const_assert_eq!(size_of::<TokenConditionalSwap>() % 8, 0);
+
+impl Default for TokenConditionalSwap {
+    fn default() -> Self {
+        Self {
+            id: 0,
+            max_buy: 0,
+            max_sell: 0,
+            bought: 0,
+            sold: 0,
+            expiry_timestamp: u64::MAX,
+            price_lower_limit: 0.0,
+            price_upper_limit: 0.0,
+            price_premium_bps: 0,
+            taker_fee_bps: 0,
+            maker_fee_bps: 0,
+            buy_token_index: TokenIndex::MAX,
+            sell_token_index: TokenIndex::MAX,
+            has_data: 0,
+            allow_creating_borrows: 0,
+            allow_creating_deposits: 0,
+            reserved: [0; 115],
+        }
+    }
+}
+
+impl TokenConditionalSwap {
+    /// Whether the entry is in use
+    ///
+    /// Note that it's possible for an entry to be in use but be expired
+    pub fn has_data(&self) -> bool {
+        self.has_data == 1
+    }
+
+    pub fn set_has_data(&mut self, has_data: bool) {
+        self.has_data = u8::from(has_data);
+    }
+
+    pub fn is_expired(&self, now_ts: u64) -> bool {
+        now_ts >= self.expiry_timestamp
+    }
+
+    pub fn allow_creating_deposits(&self) -> bool {
+        self.allow_creating_deposits == 1
+    }
+
+    pub fn allow_creating_borrows(&self) -> bool {
+        self.allow_creating_borrows == 1
+    }
+
+    pub fn remaining_buy(&self) -> u64 {
+        self.max_buy - self.bought
+    }
+
+    pub fn remaining_sell(&self) -> u64 {
+        self.max_sell - self.sold
+    }
+
+    /// Base price adjusted for the premium
+    ///
+    /// Base price is the amount of sell_token to pay for one buy_token.
+    pub fn premium_price(&self, base_price: f32) -> f32 {
+        base_price * (1.0 + self.price_premium_bps as f32 * 0.0001)
+    }
+
+    /// Premium price adjusted for the maker fee
+    pub fn maker_price(&self, premium_price: f32) -> f32 {
+        premium_price * (1.0 + self.maker_fee_bps as f32 * 0.0001)
+    }
+
+    /// Premium price adjusted for the taker fee
+    pub fn taker_price(&self, premium_price: f32) -> f32 {
+        premium_price * (1.0 - self.taker_fee_bps as f32 * 0.0001)
```

Comment:
it'd be more efficient to just store the premium and fees as f32 in the first place - and not as bps